python/dgl/data/molecule.py

```diff
@@ -0,0 +1,471 @@
+            else:
+                task_pos_weights.append(1)
+
+        self.task_pos_weights = task_pos_weights
+
+    def split_dataset(self, frac_train, frac_val, frac_test):
+        """Split the dataset into three consecutive subsets for training,
+        validation and test.
+
+        Parameters
+        ----------
+        frac_train : float
+            Proportion of data to use for training
+        frac_val : float
+            Proportion of data to use for validation
+        frac_test : float
+            Proportion of data to use for test
+        """
+        frac_total = frac_train + frac_val + frac_test
+        assert frac_total == 1, \
+            'Expect frac_train + frac_val + frac_test = 1, got {:.4f}'.format(frac_total)
+        num_datapoints = self.num_datapoints
+        train_cutoff = int(frac_train * num_datapoints)
+        val_cutoff = int((frac_train + frac_val) * num_datapoints)
+
+        total_indices = range(num_datapoints)
+        train_indices = total_indices[:train_cutoff]
+        val_indices = total_indices[train_cutoff:val_cutoff]
+        test_indices = total_indices[val_cutoff:]
+
+        self.train_set = MolSubset(train_indices, self.graphs, self.labels, self.w)
+        self.val_set = MolSubset(val_indices, self.graphs, self.labels, self.w)
+        self.test_set = MolSubset(test_indices, self.graphs, self.labels, self.w)
+
+class MolSubset(object):
+    def __init__(self, subset_indices, graphs, labels, weights):
+        """Create a subset of the dataset based on datapoint indices.
+
+        Parameters
+        ----------
+        subset_indices : list of int
+            List of datapoint indices to create the subset.
+        graphs : list of DGLGraphs
+            List of DGLGraphs for all datapoints
+        labels : np.ndarray of shape (num_datapoints, num_tasks)
+            Binary labels for all datapoints and tasks
+        weights : np.ndarray of shape (num_datapoints, num_tasks)
+            Weighting for each label and datapoint in loss computation
+        """
+        super(MolSubset, self).__init__()
+        self.subset_indices = subset_indices
+        self.graphs = [graphs[i] for i in subset_indices]
+        self.labels = F.zerocopy_from_numpy(labels[subset_indices].astype(np.float32))
+        self.weights = F.zerocopy_from_numpy(weights[subset_indices].astype(np.float32))
+
+    def __len__(self):
+        return len(self.graphs)
+
+    def __getitem__(self, item):
+        """Get the ith datapoint in the subset.
+
+        Parameters
+        ----------
+        item : int
+            Datapoint index in the subset
+
+        Returns
+        -------
+        DGLGraph
+            DGLGraph of the datapoint
+        Tensor of dtype float32
+            Labels of the datapoint for all tasks
+        Tensor of dtype float32
+            Weights of the datapoint for all tasks
+        """
+        return self.graphs[item], self.labels[item], self.weights[item]
```

Comment:
Personally I think we should follow the design of pytorch dataloader/dataset. A `Dataset` implements `_getitem_`, `torch.utils.data.Subset(dataset, indices)` to get train/test/val set, and `WeightedRandomSampler` if we need to rebalance the data.

Reply:
I've adapted PyTorch's `Subset`, but the training/validation/test set are still split during initialization as dataset splits can be very different with molecules.